pkg/cloud/services/networking/port.go

```diff
@@ -290,17 +290,32 @@ func (s *Service) DeletePorts(openStackCluster *infrav1.OpenStackCluster) error
 }
 
 func (s *Service) GarbageCollectErrorInstancesPort(eventObject runtime.Object, instanceName string) error {
-	portList, err := s.client.ListPort(ports.ListOpts{
-		Name: instanceName,
-	})
+	// NOTE: when creating an instance, ports are named using getPortName(), which
+	// features an index, so we must list all ports and delete those that start with
+	// the instance name.  Specifying the name in the list options will only return
+	// a direct match.
+	portList, err := s.client.ListPort(ports.ListOpts{})
```

Comment:
That can be a really long list to go through. Could we reasonably use the API that allows us to filter using multiple name parameters [1]? Feels like not, because we don't know the number of ports, but I'm desperately trying to find something to filter on here.

Or maybe we're able to filter on description? [2] as it's populated with a cluster name [3]?

[1] https://docs.openstack.org/api-ref/network/v2/index.html#filtering-and-column-selection
[2] https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/3d1149683d05af6980c7b745f0f9979d57266e1e/pkg/cloud/services/networking/port.go#L137
[3] https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/3d1149683d05af6980c7b745f0f9979d57266e1e/pkg/cloud/services/networking/port.go#L75-L78

Reply:
Could we infer the number of ports?  Possibly, you can generate the port from the cluster and instance spec...

I think the problem with the description filtering is it _may_ contain the cluster name, looks like you can override this at the spec level 🤷🏻

Reply:
Hm, indeed, you're right. Then it feels like generating the expected names from the spec is the safest way to go here, especially as you can override the index using `nameSuffix` [1] Could you see if the code comes out very complicated if that's attempted?

[1] https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/6912766099b348fb520ebc20e167bba72c00fb5d/pkg/cloud/services/compute/instance.go#L384-L386